platform/packages/versioning/src/release.rs

```diff
@@ -1,279 +1,268 @@
+    const ID: &'static str = env!(
+        "SOFTWARE_RELEASE_ID",
+        "No software release identifier provided as an environment variable! Please set \
+        \"SOFTWARE_RELEASE_ID\" environment variable!",
     );
 
-    const DEV_RELEASE: &'static str = "dev-release";
+    //TODO delete once deliver a version with contracts that take the current release as input
+    const PREV: &'static str = "v0.7.6";
+
+    const DEV: &'static str = "dev-release";
+
+    const VOID: &'static str = "void-release";
+
+    fn this() -> Self {
+        Self(Self::ID.into())
+    }
+
+    fn prev() -> Self {
+        Self(Self::PREV.into())
+    }
 
-    const VOID_RELEASE: &'static str = "void-release";
+    fn dev() -> Self {
+        Self(Self::DEV.into())
+    }
 
+    fn void() -> Self {
+        Self(Self::VOID.into())
+    }
+}
+
+impl Release {
     pub fn void() -> Self {
-        Self::instance(Self::VOID_RELEASE)
+        Self::instance(Id::void(), Version::new(0, SemVer::parse("0.0.0")))
     }
 
-    pub(crate) fn from_env() -> Self {
-        Self::instance(Self::RELEASE_LABEL)
+    pub(crate) fn this(code: Version) -> Self {
+        Self::instance(Id::this(), code)
     }
 
-    fn instance<L>(label: L) -> Self
-    where
-        L: Into<String>,
-    {
-        Self(label.into())
+    pub(crate) fn pull_prev(storage: &mut dyn Storage) -> Result<Self, StdError> {
+        const VERSION_STORAGE_KEY: Item<Version> = Item::new("contract_version");
+
+        VERSION_STORAGE_KEY
+            .load(storage)
+            .inspect(|_| VERSION_STORAGE_KEY.remove(storage))
+            .map(|code| Self::instance(Id::prev(), code))
+    }
+
+    fn instance(id: Id, code: Version) -> Self {
+        Self { id, code }
     }
 
-    pub fn allow_software_update(&self, current: &Version, new: &Version) -> Result<(), StdError> {
-        check_storage_match(
-            current.storage,
-            new.storage,
-            "the new software storage version",
-        )?;
+    pub fn update_software(self, to: Self) -> Result<Self, StdError> {
+        self.check_storage_match(to.code)
+            .and_then(|()| self.allow_software_update_int(to))
+    }
 
-        self.allow_software_update_int(current, new)
+    pub fn update_software_and_storage(self, to: Self) -> Result<Self, StdError> {
+        self.check_storage_adjacent(to.code)
+            .and_then(|()| self.allow_software_update_int(to))
     }
 
-    pub fn allow_software_and_storage_update<const FROM_STORAGE_VERSION: VersionSegment>(
-        &self,
-        current: &Version,
-        new: &Version,
-    ) -> Result<(), StdError> {
-        check_storage_match(
-            current.storage,
-            FROM_STORAGE_VERSION,
-            "the expected origin storage version",
-        )?;
-
-        if current.storage.wrapping_add(1) == new.storage {
-            self.allow_software_update_int(current, new)
+    fn allow_software_update_int(&self, new: Self) -> Result<Self, StdError> {
```

Comment:
`allow_` implies a state change, e.g. like a flag, rather that validation, which is what the implementation does.